### PR TITLE
rol and ror expressions

### PR DESCRIPTION
**Pull request by kroening**
- [X] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/solvers/smt2/smt2_conv.cpp [hunk 1]
```diff
@@ -1362,6 +1362,46 @@ void smt2_convt::convert_expr(const exprt &expr)
...
+
+      auto distance_int_op = numeric_cast<mp_integer>(shift_expr.distance());
+
+      if(distance_int_op.has_value())
+      {
+        out << distance_int_op.value();
+      }
+      else
+        UNEXPECTEDCASE(
+          "distance type for " + shift_expr.id_string() + "must be constant");
```
Comment (martin-cs) [inline src/solvers/smt2/smt2_conv.cpp:1393]: Can't this be implemented as a pair of shifts and an or?

File: src/util/std_expr.h [hunk 2]
```diff
@@ -2539,7 +2538,8 @@ class shift_exprt:public binary_exprt
 template <>
 inline bool can_cast_expr<shift_exprt>(const exprt &base)
 {
-  return base.id() == ID_shl || base.id() == ID_ashr || base.id() == ID_lshr;
+  return base.id() == ID_shl || base.id() == ID_ashr || base.id() == ID_lshr ||
+         base.id() == ID_ror || base.id() == ID_rol;
```
Comment (martin-cs) [inline src/util/std_expr.h:2542]: There are `lshr_exprt` et al. so it would be good to have `ror_exprt` and `rol_exprt` for consistency.

**Review by martin-cs — COMMENTED** [review]
What is here looks right (it would be nice to handle the roll-by-variable case) but what worries me is that not only is there no test, there is also no way to test as none of the front-ends actually generate these.

**Comment by tautschnig** [comment]
> What is here looks right (it would be nice to handle the roll-by-variable case) but what worries me is that not only is there no test, there is also no way to test as none of the front-ends actually generate these.

Perhaps we could have a unit test?

**Comment by kroening** [comment]
> Perhaps we could have a unit test?

That would introduce a dependency on an SMT solver, which could be quite inconvenient. And a dependency like that doesn't really fit the idea of a unit test.

The best I can think of is adding `__CPROVER_rotate_left` and `__CPROVER_rotate_right`.

**Comment by tautschnig** [comment]
> > Perhaps we could have a unit test?
> 
> That would introduce a dependency on an SMT solver, which could be quite inconvenient. And a dependency like that doesn't really fit the idea of a unit test.
> 
> The best I can think of is adding `__CPROVER_rotate_left` and `__CPROVER_rotate_right`.

I don't think we'd have to go as far as that? The unit test could generate a goto program including rotate expressions, could trigger symbolic execution, and could then request an SMT-LIB 2 output file to be generated.

**Comment by martin-cs** [comment]
On Wed, 2021-01-13 at 13:05 -0800, Michael Tautschnig wrote:
> > > Perhaps we could have a unit test?
> > 
> > That would introduce a dependency on an SMT solver, which could be
> > quite inconvenient. And a dependency like that doesn't really fit
> > the idea of a unit test.
> > 
> > The best I can think of is adding `__CPROVER_rotate_left` and
> > `__CPROVER_rotate_right`.
> 
> I don't think we'd have to go as far as that? The unit test could
> generate a goto program including rotate expressions, could trigger
> symbolic execution, and could then request an SMT-LIB 2 output file
> to be generated.

Either would be good.  Without something actually using rolls it feels
like it could very easily bit-rot.

Cheers,
 - Martin

**Comment by tautschnig** [comment]
> > Perhaps we could have a unit test?
> 
> That would introduce a dependency on an SMT solver, which could be quite inconvenient. And a dependency like that doesn't really fit the idea of a unit test.
> 
> The best I can think of is adding `__CPROVER_rotate_left` and `__CPROVER_rotate_right`.

How about us adding support for Clang's `__builtin_rotate*` intrinsics (cf. https://releases.llvm.org/8.0.0/tools/clang/docs/LanguageExtensions.html)?

**Review by tautschnig — APPROVED** [review]

**Comment by tautschnig** [comment]
@kroening I'm not entirely sure why only a single one of our regression tests is failing (check-macos-10_15-cmake-clang), but that does lead to two observations:

1. The following warnings are surprising:
```
file rotate.c line 42 function check_left64: function '__builtin_rotateleft64' is not declared
file rotate.c line 51 function check_right8: function '__builtin_rotateright8' is not declared
file rotate.c line 60 function check_right16: function '__builtin_rotateright16' is not declared
file rotate.c line 69 function check_right32: function '__builtin_rotateright32' is not declared
file rotate.c line 78 function check_right64: function '__builtin_rotateright64' is not declared
```
2. It actually because of
```
Running SMT2 QF_AUFBV using CPROVER SMT2
SMT2 solver returned error message:
	"line 107: expected ')' after rotate_left index"
```
